src/ast/nodes/shared/Node.ts

```diff
@@ -223,28 +228,52 @@ export class NodeBase implements ExpressionNode {
 		return location;
 	}
 
-	reassignPath(_path: ObjectPath, _options: ExecutionPathOptions) {}
-
-	render(code: MagicString, options: RenderOptions) {
-		this.eachChild(child => child.render(code, options));
+	parseNode(esTreeNode: GenericEsTreeNode, nodeConstructors: { [p: string]: typeof NodeBase }) {
+		for (const key of Object.keys(esTreeNode)) {
+			// That way, we can override this function to add custom initialisation and then call super.parseNode
+			if (this.hasOwnProperty(key)) continue;
+			const value = esTreeNode[key];
+			if (typeof value !== 'object' || value === null) {
+				(<GenericEsTreeNode>this)[key] = value;
+			} else if (Array.isArray(value)) {
+				(<GenericEsTreeNode>this)[key] = [];
+				for (const child of value) {
+					(<GenericEsTreeNode>this)[key].push(
+						child &&
```

Comment:
Won't this mean that there need to be checks for `boolean` wherever `this[key]` is looped over? Instead, if the `if (child)` check is moved out of the `push`, won't the array structure remain consistent with the type of the items? Was it already considered?

Reply:
This was considered. The situation is e.g. an array pattern with wholes, i.e.

```javascript
const [,a,,b] = [1, 2, 3, 4]
````

Acorn will represent this as `[null, Identifier, null, Identifier]`. The check will make sure that if the child is `null`, then `null` will be pushed, otherwise a new `Identifier` instance will be pushed. `null` is in fact the only non-entity value I am aware of that can occur in Acorn ESTree node array fields.